datadog_checks_base/tests/base/utils/test_tracing.py

```diff
@@ -18,36 +19,79 @@ def __init__(self, *args, **kwargs):
         self.instances = args[2]
         self.check_id = ''
 
+    def run(self):
+        self.check(self.instances[0])
+
+    def check(self, instance):
+        raise NotImplementedError
+
     def gauge(self, name, value):
         aggregator.submit_metric(self, self.check_id, aggregator.GAUGE, name, value, [], 'hostname', False)
 
 
 class DummyCheck(MockAgentCheck):
     def __init__(self, *args, **kwargs):
         super(DummyCheck, self).__init__(*args, **kwargs)
-        self.checked = False
 
-    def check(self, instance):
+    def check(self, _):
         self.gauge('dummy.metric', 10)
+        self.dummy_method()
+
+    def dummy_method(self):
+        self.gauge('foo', 10)
+
+
+@contextmanager
+def traced_mock_classes():
+    # temporarily unset the DDEV_TRACE_ENABLED flag to enable testing
+    # the exhaustive tracing option
+    ddev_tracce_enabled = os.environ.pop('DDEV_TRACE_ENABLED', None)
+    # the mock classes must be traced within a single test execution in order for
+    # us to be able to control the parameters used by the traced_class wrapper
+    # The regular AgentCheck common base class applies traced_class but here
+    # in the test to keep things simple we trace both base and child classes directly
+    global MockAgentCheck, DummyCheck
+    orig_mock_agent, orig_dummy = MockAgentCheck, DummyCheck
+    MockAgentCheck, DummyCheck = traced_class(MockAgentCheck), traced_class(DummyCheck)
+    yield
+    MockAgentCheck, DummyCheck = orig_mock_agent, orig_dummy
+    if ddev_tracce_enabled:
+        os.environ['DDEV_TRACE_ENABLED'] = ddev_tracce_enabled
+
+
+@pytest.mark.parametrize(
+    'integration_tracing', [pytest.param(False, id="tracing_false"), pytest.param(True, id="tracing_true")]
```

Comment:
nit: should we include a case where integration_tracing is true but exhaustive is not?

Reply:
Yes, we're already testing all the combinations of the two parameters `integration_tracing` and `integration_tracing_exhaustive:

```
tests/base/utils/test_tracing.py::test_traced_class[exhaustive_false-tracing_false] PASSED
tests/base/utils/test_tracing.py::test_traced_class[exhaustive_false-tracing_true] PASSED
tests/base/utils/test_tracing.py::test_traced_class[exhaustive_true-tracing_false] PASSED
tests/base/utils/test_tracing.py::test_traced_class[exhaustive_true-tracing_true] PASSED
```

Reply:
ok wow did not realize parametrize behaved like this, amazing 👍